src/main/java/jcifs/smb1/ntlmssp/Type2Message.java

```diff
@@ -0,0 +1,368 @@
+     */ 
+    public byte[] getTargetInformation() {
+        return targetInformation;
+    }
+
+    /**
+     * Sets the target information block.
+     * The target information block is used by the client to create
+     * an NTLMv2 response.
+     * 
+     * @param targetInformation The target information block.
+     */
+    public void setTargetInformation(byte[] targetInformation) {
+        this.targetInformation = targetInformation;
+    }
+
+    /**
+     * Returns the local security context.
+     *
+     * @return A <code>byte[]</code> containing the local security
+     * context.  This is used by the client to negotiate local
+     * authentication.
+     */
+    public byte[] getContext() {
+        return context;
+    }
+
+    /**
+     * Sets the local security context.  This is used by the client
+     * to negotiate local authentication.
+     *
+     * @param context The local security context.
+     */
+    public void setContext(byte[] context) {
+        this.context = context;
+    }
+
+    public byte[] toByteArray() {
+        try {
+            String targetName = getTarget();
+            byte[] challenge = getChallenge();
+            byte[] context = getContext();
+            byte[] targetInformation = getTargetInformation();
+            int flags = getFlags();
+            byte[] target = new byte[0];
+            if ((flags & NTLMSSP_REQUEST_TARGET) != 0) {
+                if (targetName != null && targetName.length() != 0) {
+                    target = (flags & NTLMSSP_NEGOTIATE_UNICODE) != 0 ?
+                            targetName.getBytes(UNI_ENCODING) :
+                            targetName.toUpperCase().getBytes(getOEMEncoding());
+                } else {
+                    flags &= (0xffffffff ^ NTLMSSP_REQUEST_TARGET);
+                }
+            }
+            if (targetInformation != null) {
+                flags |= NTLMSSP_NEGOTIATE_TARGET_INFO;
+                // empty context is needed for padding when t.i. is supplied.
+                if (context == null) context = new byte[8];
+            }
+            int data = 32;
+            if (context != null) data += 8;
+            if (targetInformation != null) data += 8;
+            byte[] type2 = new byte[data + target.length +
+                    (targetInformation != null ? targetInformation.length : 0)];
+            System.arraycopy(NTLMSSP_SIGNATURE, 0, type2, 0, 8);
+            writeULong(type2, 8, 2);
+            writeSecurityBuffer(type2, 12, data, target);
+            writeULong(type2, 20, flags);
+            System.arraycopy(challenge != null ? challenge : new byte[8], 0,
+                    type2, 24, 8);
+            if (context != null) System.arraycopy(context, 0, type2, 32, 8);
+            if (targetInformation != null) {
+                writeSecurityBuffer(type2, 40, data + target.length,
+                        targetInformation);
+            }
+            return type2;
+        } catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)